lib/improver/cli/wxcode.py

```diff
@@ -31,123 +31,30 @@
+def _extend_help(fn):
+    # TODO: speed up help - pulling in decision tree imports iris
+    # (and gets executed at import time)
+    from improver.wxcode.wxcode_utilities import interrogate_decision_tree
+    for wxtree in ('high_resolution', 'global'):
+        title = wxtree.title().replace('_', ' ') + ' tree inputs'
+        inputs = interrogate_decision_tree(wxtree).replace('\n', '\n    ')
+        tree_help = f"""
+{title}:
 
-
-def interrogate_decision_tree(wxtree):
-    """
-    Obtain a list of necessary inputs from the decision tree as it is currently
-    defined. Return a list of the diagnostic names, the thresholds needed,
-    and whether they are thresholded above or below these values. This output
-    is used to create the CLI help, informing the user of the necessary inputs.
-
-    Returns:
-        list of str:
-            Returns a list of strings, an entry for each diagnostic required,
-            including threshold details.
-    """
-
-    # Get current weather symbol decision tree and populate a list of
-    # required inputs for printing.
-    if wxtree == 'high_resolution':
-        queries = wxcode_decision_tree()
-    elif wxtree == 'global':
-        queries = wxcode_decision_tree_global()
-
-    # Diagnostic names and threshold values.
-    requirements = {}
-    # How the data has been thresholded relative to these thresholds.
-    relative = {}
-
-    for query in queries.values():
-        diagnostics = expand_nested_lists(query, 'diagnostic_fields')
-        for index, diagnostic in enumerate(diagnostics):
-            if diagnostic not in requirements:
-                requirements[diagnostic] = []
-                relative[diagnostic] = []
-            requirements[diagnostic].extend(
-                expand_nested_lists(query, 'diagnostic_thresholds')[index])
-            relative[diagnostic].append(
-                expand_nested_lists(query, 'diagnostic_conditions')[index])
-
-    # Create a list of formatted strings that will be printed as part of the
-    # CLI help.
-    output = []
-    for requirement in requirements:
-        entries = np.array([entry for entry in requirements[requirement]])
-        relations = np.array([entry for entry in relative[requirement]])
-        _, thresholds = np.unique(np.array([item.points.item()
-                                            for item in entries]),
-                                  return_index=True)
-        output.append('{}; thresholds: {}'.format(
-            requirement, ', '.join([
-                '{} ({})'.format(str(threshold.points.item()),
-                                 str(threshold.units))
-                for threshold, relation in
-                zip(entries[thresholds], relations[thresholds])])))
-    return output
-
-
-def main(argv=None):
-    """Parser to accept input data and an output destination before invoking
-    the weather symbols plugin.
+    {inputs}
     """
+        fn.__doc__ += tree_help
+    return fn
```

Comment:
As this is implemented as a decorator, is it invoked every time the CLI is called, regardless of whether help is requested/cubes are provided? Seems like an unnecessary overhead for the latter as it will be called every time the CLI is used in the suite.

Reply:
Yep - from a quick test it adds about ~1s onto the runtime required to generate the help (so probably 1s on top of everything else as well). I don't think it's necessarily anything to be too concerned about.